_posts/2021-03-15-wildfly-clustering-spring-session.adoc

```diff
@@ -0,0 +1,115 @@
+Infinispan's existing Spring Session implementation stores instances of `org.springframework.session.MapSession` into a Spring cache, which is backed by either an embedded or remote cache.
+
+However, like many `MapSession` based `SessionRepository` implementations, it suffers from a common set of problems/limitations:
+
+* The `MapSession` object, including all session metadata and attributes must be serialized and replicated/persisted on every request, regardless of the extent to which the session was modified.
+** Because Spring Session triggers session persistence from a `ServletFilter`, this cost is borne per request.
+* Since each request for a given session operates on a copy of the `MapSession`, and since the `MapSession` is not persisted/replicated until the end of the request:
+** Session modifications in one requests are not visible to concurrent requests for the same session.
+** A given session modification can be overwritten by a previous modification of the same session from a concurrent request if that request finishes first.
+* Session attributes implementing `HttpSessionActivationListener` are never notified before/after they are marshalled/unmarshalled.
+
+== How does it work?
+
+wildfly-clustering-spring-session is implemented as a Spring `SessionRepository` facade to a container-independent `SessionManager` implementation that persists session attributes and metadata to either an embedded Infinispan cache or a remote Infinispan server cluster via HotRod.
+This grants all of the features to Spring Session that one would normally expect when deploying a distributed web application to WildFly, including:
+
+* Minimal persistence payloads using configurable persistence granularity.
+* Consistent view of a given session from multiple concurrent request threads on a given server
+* Mutable session attribute semantics (emulates the semantics of local session manager)
+** Ability to treat specific session attributes as immutable
+* Configurable limit for the number of active sessions to retain in memory
+* Proper `HttpSessionActivationListener` notifications.
+* Configurable session attribute marshaller (ProtoStream, JBoss Marshalling, or Java Serialization)
+
+
+== How do I use it?
+
+In this example, we'll use a remote Infinispan cluster containing a single member, deployed to the localhost, using its default configuration.
+See the project https://github.com/wildfly-clustering/wildfly-clustering-spring-session/blob/master/README.md[README] for details on the available configuration annotations and their attributes.
+
+1. Create a class containing the session repository configuration.
+e.g.
++
+[source,java]
+----
+@org.wildfly.clustering.web.spring.hotrod.annotation.EnableHotRodHttpSession(@HotRod(uri = "hotrod://127.0.0.1:11222"), manager = @SessionManager(granularity = ATTRIBUTE))
+public class Config {
+	// ...
+}
+----
+
+1. Ensure your `/WEB-INF/applicationContext.xml` will process you configuration annotation:
```

Comment:
```suggestion
2. Ensure your `/WEB-INF/applicationContext.xml` will process you configuration annotation:
```

Reply:
@oraNod I thought asciidoc handles the numbering during rendering, no?

Reply:
It's quite handy, actually.  e.g. if you add a step in the middle, you don't need to re-number!